src/mod/endpoints/mod_gsmopen/gsmlib/gsmlib-1.10-patched-13ubuntu/doc/gsmctl.man
.TH GSMCTL 8 "##DATE##" "gsmctl v##VERSION##"
.SH NAME
gsmctl, gsmsiectl \- GSM mobile phone control program
.SH SYNOPSIS
.B gsmctl
.RB [\| \-b
.IR  baudrate \|]
.RB [\| \-\-baudrate
.IR baudrate\| ]
.RB [ \|\-d
.IR device\| ]
.RB [ \|\-\-device
.IR device\fP ]
.RB [ \|\-h\| ]
.RB [ \|\-\-help\| ]
.RB [ \|\-I
.IR "init string" \|]
.RB [ \|\-\-init
.IR "init string" \|]
.RB [ \|\-v\| ]
.RB [ \|\-\-version\| ]
.RB [ \|\-X\| ]
.RB [ \|\-\-xonxoff\| ]
.BI \-o \ operation
.RB  | \ \-\-operation
.IR operation \ |
.I parameters
.PP
.B gsmsiectl
.RB [\| \-b
.IR  baudrate \|]
.RB [\| \-\-baudrate
.IR baudrate\| ]
.RB [ \|\-d
.IR device\| ]
.RB [ \|\-\-device
.IR device\fP ]
.RB [ \|\-h\| ]
.RB [ \|\-\-help\| ]
.RB [ \|\-I
.IR "init string" \|]
.RB [ \|\-\-init
.IR "init string" \|]
.RB [ \|\-v\| ]
.RB [ \|\-\-version\| ]
.RB [ \|\-X\| ]
.RB [ \|\-\-xonxoff\| ]
.BI \-o \ operation
.RB  | \ \-\-operation
.IR operation \ |
.I parameters
.SH DESCRIPTION
.B gsmctl
can request information from or perform operations on an GSM mobile
phone.
.PP
.B gsmctl 
attaches itself to the 
.I device
given on the command line (usually an GSM modem) using the specified
.IR baudrate .  
If no
.I device
is given, the device
.I /dev/mobilephone
is used. If no 
.I baudrate
is given, a default baud rate of 38400 is used.
.PP
.B gsmctl
can be used in to modes: If no 
.B \-\-operation
option is given the
.I parameters
specify the status information to be retrieved from
the mobile phone. See the section 
.B STATUS INFORMATION
for more details. If an 
.B \-\-operation
option is given the requested
.I operation
is performed on the mobile using the
.IR parameters . 
See the section 
.B OPERATIONS
for more details.
.PP
.B gsmsiectl
is the same program with some extension for Siemens mobile phones.
Some extra 
.B OPERATIONS
are available in this case.
.PP
Error messages are printed to the standard error output.  If the
program terminates on error the error code 1 is returned.
.SH OPTIONS
.TP
.BI \-b\  baudrate ,\ \-\-baudrate\  baudrate
The baud rate to use.  Defaults to 38400.
.TP
.BI \-d\  device ,\ \-\-device\  device
The device to which the GSM modem is connected. The default is
.IR /dev/mobilephone .
.TP
.B \-h,\ \-\-help
Prints an option summary.
.TP
.BI \-I\  "init string" ,\ \-\-init\  "init string"
Initialization string to send to the TA (default: "E0"). Note that the
sequence "ATZ" is sent first.
.TP
.BI \-o\  operation ,\ \-\-operation\  operation
This option is used to perform an operation on the mobile phone. Refer
to the section 
.B OPERATIONS
for more information on the available
operations and the 
.I parameters
required for each operation.
.TP
.B \-v,\ \-\-version
Prints the program version.
.TP
.B \-X,\ \-\-xonxoff
Uses software handshaking (XON/XOFF) for accessing the device.
.SH STATUS INFORMATION
If called without the 
.B \-\-operation
option 
.B gsmctl
prints out default mobile phone status information as specified by the
.I parameters
described below:
.TP 7
.B ALL
Prints all available information.
.TP 7
.B BATT
Prints out information about the current battery status. Two lines of
the form "<BATT0> text" and "<BATT1>  charge" are printed. Text may be 
of "0 ME is powered by the battery", "1 ME has a battery connected,
but is not powered by it", "2 ME does not have a battery connected",
or  "3 Recognized power fault, calls inhibited". Charge is a number in 
the range 0..100 where 0 means that the battery is empty or not
connected and 100 means full charge.
.TP 7
.B BITERR
Prints information about the current bit error rate. The output is of
form "<BITERR0>  value" where value can be of 0..7, 99 (99 means not
available or not detectable).
.TP 7
.B CLIP
This option prints a line of the form "<CLIP0> on" if caller line
identification is turned on in the network, "<CLIP0> off" otherwise.
.TP 7
.B CSET
Display info about charsets.  The output is in the form "<CSET0>
available" and "<CSET1> current", which describes the charsets
available and which is current.
.TP 7
.B CURROP
Prints information about the current operator. The output line has the
form "<CURROP0> Long name: \'xxxx\' Short name: \'yyyy\' Numeric Name: zzzz".
.TP 7
.B FLCAP
Prints out the two-letter names of available facility locks. The
meaning of standardized facility lock names is as follows:
.RS
.TP 3
.I CS
Lock control surface (eg. phone keyboard)
.TP 3
.I PS
Lock phone to SIM card (mobile phone asks password when other than
current SIM card inserted)
.TP 3
.I SC
Lock SIM card (SIM asks password in mobile phone power-up and when
this lock command issued)
.TP 3
.I AO
Barr all outgoing calls
.TP 3
.I OI
Barr outgoing international calls
.TP 3
.I OX
Barr outgoing international calls except to home country
.TP 3
.I AI
Barr all incoming calls
.TP 3
.I IR
Barr incoming calls when roaming outside the home country
.TP 3
.I NT
Barr incoming calls from numbers not stored to TA memory
.TP 3
.I NM
Barr incoming calls from numbers not stored to mobile phone memory
.TP 3
.I NS
Barr incoming calls from numbers not stored to SIM memory
.TP 3
.I NA
Barr incoming calls from numbers not stored in Any memory
.TP 3
.I AB
All Barring services (FIXME)
.TP 3
.I AG
All outGoing barring services (FIXME)
.TP 3
.I AC
All inComing barring services (FIXME)
.TP 3
.I FD
SIM fixed dialling memory feature (if PIN2 authentication has not been
done during the current session, PIN2 is required as <passwd>)
.PP
Note that mobile phones may implement facility locks not documented
here.
.RE
.TP 7
.B FLSTAT
Prints information about the status of facility locking in the mobile
phone. For each facility a line of the form "<FLSTATn> \'facilityname\'
classes" is printed. The classes may be "Voice", "Data", and/or "Fax"
depending on the class for which the facility is enabled. If "unknown"
is printed out this means that the TA is not able to report the
status. If the facility is not enabled for any class, no class
identifier is printed. See 
.B FLCAP
for information about the available two-letter names of facility
locks.
.TP 7
.B FORW
Prints information about call forwarding enabled in the mobile
phone. For each combination of reason (UnconditionalReason,
MobileBusyReason, NoReplyReason, and NotReachableReason) and class
(Voice, Data, and Fax) a line of the form "<FORWn.m> reason class
number: xxx subaddr: yyy time: zzz" is printed. The time is the number 
of seconds to wait before forwarding in case of NoReplyReason.
.TP 7
.B ME
Prints manufacturer, model, revision, and serial number of
the mobile phone (ME = mobile equipment). Output lines are numbered from 
"<ME0>" to "<ME3>".
.TP 7
.B OP
Prints information about the currently available mobile network
operators. For each operator an output line of the form "<OPn> status 
Long name: \'xxxx\' Short name: \'yyyy\' Numeric name: zzz" is returned
where status may be of:
.RS
.TP 10
.I unknown
The status of the operator is unknown.
.TP
.I current
This is the currently selected operator.
.TP
.I available
This operator is available for selection.
.TP
.I forbidden
This operator is not available for selection, ie. not
accessible with this SIM card.
.RE
.TP 7
.B PIN
Prints information about the current PIN status. The output is of
form "<PIN0>  status". Where status can be one of the following lines:
.RS
.TP 10
.I READY
ME is not pending for any password.
.TP
.I SIM PIN
ME is waiting SIM PIN to be given.
.TP
.I SIM PUK
ME is waiting SIM PUK to be given.
.TP
.I PH\-SIM PIN
ME is waiting phone/-to/-SIM card password to be given.
.TP
.I PH/-FSIM PIN
ME is waiting phone/-to/-very first SIM card password to be given.
.TP
.I PH/-FSIM PUK
ME is waiting phone/-to/-very first SIM card unblocking password to be
given.
.TP
.I SIM PIN2
ME is waiting SIM PIN2 to be given.
.TP
.I SIM PUK2
ME is waiting SIM PUK2 to be given.
.TP
.I PH/-NET PIN
ME is waiting network personalisation password to be given.
.TP
.I PH/-NET PUK
ME is waiting network personalisation unblocking password to be given.
.TP
.I PH/-NETSUB PIN
ME is waiting network subset personalisation password to be given.
.RE
.TP 7
.B PW
Facilities in the mobile phone may be protected by passwords (ie. PINs 
or PUKs). This option prints out lines of the form "<PWn>
\'facilityname\' len" for each facility for which a facility lock
password exists. See 
.B FLCAP
for information about the available two-letter names of facility
locks.
.TP 7
.B SCA
Reports the default SMS service centre address currently set in the
mobile phone.
.TP 7
.B SIG
Prints information about the current network signal
strength. The output is of the form "<SIG0> value". Following values
are possible:
.TP
.po +7
.I 0
-113 dBm or less
.TP
.I 1
-111 dBm
.TP
.I 2...30
-109... -53 dBm (in steps of 2 dBm)
.TP
.I 31
-51 dBm or greater
.TP
.I 99
not known or not detectable
.PP
.po 
.SH OPERATIONS
.TP
.BI dial\  number
Dials 
.IR number .
After dialling 
.B gsmctl
waits for a <CR> keypress to terminate.
.PP
.B forw
.IR mode\ reason\ number\  [\| facilityclass\| ]\ [ forwardtime ]
.RS
Changes the call forwarding behaviour in the network. 
.I mode
can be any of:
.RS 3
.I disable
Call forwarding is disabled.
.PP
.I enable
Call forwarding is enabled. The call forwarding reason, number, and
(optionally) forwardtime must be registered beforehand.
.PP
.I register
This registers the call forwarding reason, number, and forwardtime in
the network.
.PP
.I erase
This is the counterpart to register. It erase the call forwarding info 
in the network.
.RE
.PP
.I reason
can be any of:
.RS 3
.I unconditional
This applies to every call.
.PP
.I mobilebusy
The mobile phone is busy.
.PP
.I noreply
The call is not answered in \fIforwardtime\fP seconds.
.PP
.I notreachable
Mobile phone cannot be reached (ie. switched off).
.PP
.I all
This refers to all forwarding reasons.
.PP
.I allconditional
This refers to all conditional forward reasons 
.RI ( mobilebusy ,\  noreply ,\  notreachable ).
.RE
.PP
.I number
is the number the incoming call is forwarded to. See the explanation
for 
.I lock
operation for the meaning of 
.IR facilityclass .

ChangeLog
 * libZRTP SDK library, implements the ZRTP secure VoIP protocol.
 * Copyright (c) 2006-2009 Philip R. Zimmermann.  All rights reserved.
 * Contact: http://philzimmermann.com
 * For licensing and other legal details, see the file zrtp_legal.c.

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
SINCE LIBZRTP v0.80 CHANGELOG IS A PART OF HTML DOCUMENTATION. 
https://developers.zfoneproject.com/libzrtp/wiki/LibzrtpChangeLog
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!


libzrtp 0.7.1											18.11.2008
--------------------------------------------------------------------------------
1.	Fixed bug with hardcoded AES128 cipher for generating SRTP keys. In this 
	version is selected according to ZRTP discovery.
	
2.	Added initialization/deinitalization functions to zrtp helper functions.

libzrtp 0.7.0											04.11.2008
--------------------------------------------------------------------------------
1.	Changes in libzrtp sources tree.

2.	Improvements in libzrtp initialization routine:
	- all global zrtp options were combined in zrtp_config_t structure;
	- zrtp_init() allocates memory for zrtp global context;
	- zrtp_config_defaults() 

3.	Improvements for Passive/Active mode support.
	a) A passive endpoint never sends a commit message, period. Also, it 
	declares itself as a passive endpoint by setting the P flag it its own
	Hello message;
	b) A active endpoint does not send a commit to a passive endpoint, which it 
	recognizes by detecting the P flag;
	c) A passive phone, if acting as a SIP initiator (meaning it initiated the 
	call), rejects all commit packets from everyone;
	d) A passive phone rejects all commit messages from a PBX, which is easily 
	recognized by the M flag.
	Passive mode support is built into the library logic and will be used
	automatically if the developer specifies signaling role by setting 
	is_initiator flag in zrtp_init_session_ctx().

4.	Improvements in ZRTP feedback interface and system-dependent functions.
    There are two types of interface functions in libzrtp: system dependent API
	and realization of helper functions and events. System dependent API in
	defined in zrtp_iface_system.h as set of extern functions. System
	functions are already implemented for several basic platforms in
	zrtp_iface.sys.c libzrtp feedback and helper functions were re-factored
	and implemented as set of callbacks. If the developer doesn't want to
	handle one or another event it may just leave necessary pointer empty.
	See zrtp_init() and zrtp_callback_t, zrtp_iface.h for more details.

5. 	ZRTP configuration approach was improved: zrtp_config_xxx.h contains
	adjustments for necessary target platform. libzrtp contains default
	configs for Linux, OS X, Windows, Window CE and Symbian platforms.
	All ZRTP protocol and behavior related adjustments are collected in
	zrtp_config_user.h. Edit this file to configure libzrtp for your
	needs.
	
6.	Implemented new functions in protocol according to the Internet Draft v 10.

7.	Improved realization of built-in libzrtp scheduler. Fixed bug with crashing
	on performing delay call when zrtp session have been already deleted.
	
8. 	Logging function was improved. Use ZRTP_LOG macro to print log messages. See
	zrtp_log.h for more information.


libzrtp 0.6.8											03.09.2008
--------------------------------------------------------------------------------
ZRTP
1.	Fixed bug with incorrect maximum value for T1 retry interval timer which 
    increased delay between LOOKING_FOR_ZRTP and NO_ZRTP_SUPPORT states. 
    Internal fix, no API changes required;
2.	Implemented version negotiation according to the latest specification.
	libzrtp v 0.6.8 supports ZRTP v0.90 only. No changes required in 
	applications that use the SDK.
3. 	Compilation flag WITH_ZFONE was removed. The developer, who wants to use
	built-in ZRTP cache, has to set name of the ZRTP cache explicitly,
	implementing  zrtp_get_cache_path() function.
4. 	New libzrtp licensing scheme was implemented. It allows the licensing policy 
    to be changed at run time.  See zrtp_license_mode_t doc. for more details. 
    Affected API - zrtp_init().
5. 	ZRTP Protocol version was changed to 0.90 according to ZRTP Internet Draft.
6. 	Some changes in Linux config files: surplus configuration flags were removed
	from ./cfg.XXX templates. 
7.	Added experimental ZRTP messages retries scheduler for slow channels. As 
	ex ample for GSM CSD channel with average bandwidth 6Kb/s. To use this
	option build library with BUILD_FOR_CSD flag.


libzrtp 0.6.6											27.06.2008
--------------------------------------------------------------------------------
ZRTP
1. Small bug was fixed in S0 calculation: when RS1 is corrupted the library uses
   RS2 instead;
2. Some changes in Makefile and building process: unused header were eliminated 	
   from the installation process.
3. -DBUILD_WITH_ZRTP_MUTEXES was replaced with --enable_mutexes option passed to 

src/mod/endpoints/mod_gsmopen/gsmlib/gsmlib-1.10-patched-13ubuntu/apps/gsmctl.cc
// *************************************************************************
// * GSM TA/ME library
// *
// * File:    gsmctl.cc
// *
// * Purpose: GSM mobile phone control program
// *
// * Author:  Peter Hofmann (software@pxh.de)
// *
// * Created: 11.7.1999
// *************************************************************************

#ifdef HAVE_CONFIG_H
#include <gsm_config.h>
#endif
#include <gsmlib/gsm_nls.h>
#include <string>
#if defined(HAVE_GETOPT_LONG) || defined(WIN32)
#include <getopt.h>
#endif
#include <strstream>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <gsmlib/gsm_me_ta.h>
#include <gsmlib/gsm_util.h>
#include <gsmlib/gsm_sysdep.h>
#ifdef WIN32
#include <gsmlib/gsm_win32_serial.h>
#else
#include <gsmlib/gsm_unix_serial.h>
#include <unistd.h>
#endif
#include <iostream>

using namespace std;
using namespace gsmlib;

// my ME

static MeTa *m;

// information parameters

enum InfoParameter {AllInfo, // print all info
                    MeInfo,     // MeInfo must be first!
                    FunctionalityInfo,
                    OperatorInfo,
                    CurrentOperatorInfo,
                    FacilityLockStateInfo,
                    FacilityLockCapabilityInfo,
                    PasswordInfo,
                    PINInfo,
                    CLIPInfo,
                    CallForwardingInfo,
                    BatteryInfo,
                    BitErrorInfo,
                    SCAInfo,
                    CharSetInfo,
                    SignalInfo}; // SignalInfo must be last!

// operation parameters

// FIXME operations not implemented yet

// options

#ifdef HAVE_GETOPT_LONG
static struct option longOpts[] =
{
  {"xonxoff", no_argument, (int*)NULL, 'X'},
  {"operation", required_argument, (int*)NULL, 'o'},
  {"device", required_argument, (int*)NULL, 'd'},
  {"baudrate", required_argument, (int*)NULL, 'b'},
  {"init", required_argument, (int*)NULL, 'I'},
  {"help", no_argument, (int*)NULL, 'h'},
  {"version", no_argument, (int*)NULL, 'v'},
  {(char*)NULL, 0, (int*)NULL, 0}
};
#else
#define getopt_long(argc, argv, options, longopts, indexptr) \
  getopt(argc, argv, options)
#endif

// helper function, prints forwarding info

void printForwardReason(string s, ForwardInfo &info)
{
  cout << s << "  "
       << (info._active ? _("active ") : _("inactive "))
       << _("number: ") << info._number
       << _("  subaddr: ") << info._subAddr
       << _("  time: ") << info._time << endl;
}

// print information

static void printInfo(InfoParameter ip)
{
  switch (ip)
  {
  case MeInfo:
  {
    MEInfo mei = m->getMEInfo();
    cout << _("<ME0>  Manufacturer: ") << mei._manufacturer << endl
         << _("<ME1>  Model: ") << mei._model << endl
         << _("<ME2>  Revision: ") << mei._revision << endl
         << _("<ME3>  Serial Number: ") << mei._serialNumber << endl;
    break;
  }
  case FunctionalityInfo:
  {
    try {
      int fun;
      fun = m->getFunctionalityLevel();
      cout << _("<FUN>  Functionality Level: ") << fun << endl;
    } catch (GsmException &x) { 
      cout << _("<FUN>  Functionality Level: ") << _("unsupported") << endl;
    }
    break;
  }
  case OperatorInfo:
  {
    int count = 0;
    vector<OPInfo> opis = m->getAvailableOPInfo();
    for (vector<OPInfo>::iterator i = opis.begin(); i != opis.end(); ++i)
    {
      cout << "<OP" << count << _(">  Status: ");
      switch (i->_status)
      {
      case UnknownOPStatus: cout << _("unknown"); break;
      case CurrentOPStatus: cout << _("current"); break;
      case AvailableOPStatus: cout << _("available"); break;
      case ForbiddenOPStatus: cout << _("forbidden"); break;
      }
      cout << _("  Long name: '") << i->_longName << "' "
           << _("  Short name: '") << i->_shortName << "' "
           << _("  Numeric name: ") << i->_numericName << endl;
      ++count;
    }
    break;
  }
  case CurrentOperatorInfo:
  {
    OPInfo opi = m->getCurrentOPInfo();
    cout << "<CURROP0>"
         << _("  Long name: '") << opi._longName << "' "
         << _("  Short name: '") << opi._shortName << "' "
         << _("  Numeric name: ") << opi._numericName
         << _("  Mode: ");
    switch (opi._mode)
    {
    case AutomaticOPMode: cout << _("automatic"); break;
    case ManualOPMode: cout << _("manual"); break;
    case DeregisterOPMode: cout << _("deregister"); break;
    case ManualAutomaticOPMode: cout << _("manual/automatic"); break;
    }
    cout << endl;
    break;
  }
  case FacilityLockStateInfo:
  {
    int count = 0;
    vector<string> fclc = m->getFacilityLockCapabilities();
    for (vector<string>::iterator i = fclc.begin(); i != fclc.end(); ++i)
      if (*i != "AB" && *i != "AG" && *i != "AC")
      {
        cout << "<FLSTAT" << count <<  ">  '" << *i << "'";
        try
        {
          if (m->getFacilityLockStatus(*i, VoiceFacility))
            cout << _("  Voice");
        }
        catch (GsmException &e)
        {
          cout << _("  unknown");
        }
        try
        {
        if (m->getFacilityLockStatus(*i, DataFacility))
          cout << _("  Data");
        }
        catch (GsmException &e)
        {
          cout << _("  unknown");
        }
        try
        {
        if (m->getFacilityLockStatus(*i, FaxFacility))
          cout << _("  Fax");
        }
        catch (GsmException &e)
        {
          cout << _("  unknown");
        }
        cout << endl;
        ++count;
      }
    break;
  }
  case FacilityLockCapabilityInfo:
  {
    cout << "<FLCAP0>  ";
    vector<string> fclc = m->getFacilityLockCapabilities();
    for (vector<string>::iterator i = fclc.begin(); i != fclc.end(); ++i)
      cout << "'" << *i << "' ";
    cout << endl;
    break;
  }
  case PasswordInfo:
  {
    vector<PWInfo> pwi = m->getPasswords();
    int count = 0;
    for (vector<PWInfo>::iterator i = pwi.begin(); i != pwi.end(); ++i)
    {
      cout << "<PW" << count <<  ">  '"
           << i->_facility << "' " << i->_maxPasswdLen << endl;
      ++count;
    }
    break;
  }
  case PINInfo:
  {
    cout << "<PIN0> " << m->getPINStatus() << endl;
    break;
  }
  case CLIPInfo:
  {
    cout << "<CLIP0>  " << (m->getNetworkCLIP() ? _("on") : _("off")) << endl;
    break;
  }
  case CallForwardingInfo:
  {
    for (int r = 0; r < 4; ++r)
    {
      string text;
      switch (r)
      {
      case 0: text = _("UnconditionalReason"); break;
      case 1: text = _("MobileBusyReason"); break;
      case 2: text = _("NoReplyReason"); break;
      case 3: text = _("NotReachableReason"); break;
      }
      ForwardInfo voice, fax, data;
      m->getCallForwardInfo((ForwardReason)r, voice, fax, data);
      cout << "<FORW" << r << ".";
      printForwardReason("0>  " + text + _("  Voice"), voice);
      cout << "<FORW" << r << ".";
      printForwardReason("1>  " + text + _("  Data"), data);
      cout << "<FORW" << r << ".";
      printForwardReason("2>  " + text + _("  Fax"), fax);
    }
    break;
  }
  case BatteryInfo:
  {
    cout << "<BATT0>  ";
    int bcs = m->getBatteryChargeStatus();
    switch (bcs)
    {
    case 0: cout << _("0 ME is powered by the battery") << endl; break;
    case 1: cout << _("1 ME has a battery connected, but is not powered by it")
                 << endl; break;
    case 2: cout << _("2 ME does not have a battery connected") << endl; break;
    case 3:
      cout << _("3 Recognized power fault, calls inhibited") << endl;
      break;
    }
    cout << "<BATT1>  " << m->getBatteryCharge() << endl;
    break;
  }
  case BitErrorInfo:
  {
    cout << "<BITERR0>  " << m->getBitErrorRate() << endl;
    break;
  }
  case SCAInfo:
  {
    cout << "<SCA0>  " << m->getServiceCentreAddress() << endl;
    break;
  }
  case CharSetInfo:
  {
    cout << "<CSET0>  ";
    vector<string> cs = m->getSupportedCharSets();
    for (vector<string>::iterator i = cs.begin(); i != cs.end(); ++i)
      cout << "'" << *i << "' ";
    cout << endl;
    cout << "<CSET1>  '" << m->getCurrentCharSet() << "'" << endl;
    break;
  }
  case SignalInfo:
  {
    cout << "<SIG0>  " << m->getSignalStrength() << endl;
    break;
  }
  default:
    assert(0);
    break;
  }
}

// convert facility class string of the form "", "all", or any combination
// of "v" (voice), "d" (data), or "f" (fax) to numeric form

FacilityClass strToFacilityClass(string facilityClassS)
{
  facilityClassS = lowercase(facilityClassS);
  FacilityClass facilityClass = (FacilityClass)0;
  if (facilityClassS == "all" || facilityClassS == "")
    return (FacilityClass)ALL_FACILITIES;

  // OR in facility class bits
  for (unsigned int i = 0; i < facilityClassS.length(); ++i)
    if (facilityClassS[i] == 'v')
      facilityClass = (FacilityClass)(facilityClass | VoiceFacility);
    else if (facilityClassS[i] == 'd')
      facilityClass = (FacilityClass)(facilityClass | DataFacility);
    else if (facilityClassS[i] == 'f')
      facilityClass = (FacilityClass)(facilityClass | FaxFacility);
    else
      throw GsmException(
        stringPrintf(_("unknown facility class parameter '%c'"),
                     facilityClassS[i]), ParameterError);

  return facilityClass;
}

// check if argc - optind is in range min..max
// throw exception otherwise

void checkParamCount(int optind, int argc, int min, int max)
{
  int paramCount = argc - optind;
  if (paramCount < min)
    throw GsmException(stringPrintf(_("not enough parameters, minimum number "
                                      "of parameters is %d"), min),
                       ParameterError);
  else if (paramCount > max)
    throw GsmException(stringPrintf(_("too many parameters, maximum number "
                                      "of parameters is %d"), max),
                       ParameterError);
}

// *** main program

int main(int argc, char *argv[])
{
  try
  {
    // handle command line options
    string device = "/dev/mobilephone";
    string operation;
    string baudrate;
    string initString = DEFAULT_INIT_STRING;
    bool swHandshake = false;

    int opt;
    int dummy;
    while((opt = getopt_long(argc, argv, "I:o:d:b:hvX", longOpts, &dummy))
          != -1)
      switch (opt)
      {
      case 'X':
        swHandshake = true;
        break;
      case 'I':
        initString = optarg;
        break;
      case 'd':
        device = optarg;
        break;
      case 'o':
        operation = optarg;
        break;
      case 'b':
        baudrate = optarg;
        break;
      case 'v':
        cerr << argv[0] << stringPrintf(_(": version %s [compiled %s]"),
                                        VERSION, __DATE__) << endl;
        exit(0);
        break;
      case 'h':
        cerr << argv[0] << _(": [-b baudrate][-d device][-h]"
                             "[-I init string][-o operation]\n"
                             "  [-v][-X]{parameters}") << endl
             << endl
             << _("  -b, --baudrate    baudrate to use for device "
                  "(default: 38400)")
             << endl
             << _("  -d, --device      sets the destination device to "
                  "connect to") << endl
             << _("  -h, --help        prints this message") << endl
             << _("  -I, --init        device AT init sequence") << endl
             << _("  -o, --operation   operation to perform on the mobile \n"
                  "                    phone with the specified parameters")
             << endl
             << _("  -v, --version     prints version and exits") << endl
             << _("  -X, --xonxoff     switch on software handshake") << endl
             << endl
             << _("  parameters        parameters to use for the operation\n"
                  "                    (if an operation is given) or\n"
                  "                    a specification which kind of\n"
                  "                    information to read from the mobile "
                  "phone")
             << endl << endl
             << _("Refer to gsmctl(1) for details on the available parameters"
                  " and operations.")
             << endl << endl;
        exit(0);
        break;
      case '?':
        throw GsmException(_("unknown option"), ParameterError);
        break;
      }

    // open the port and ME/TA
    m = new MeTa(new
#ifdef WIN32
                 Win32SerialPort
#else
                 UnixSerialPort
#endif
                 (device,
                  baudrate == "" ?
                  DEFAULT_BAUD_RATE :
                  baudRateStrToSpeed(baudrate),
                  initString, swHandshake));
    
    if (operation == "")
    {                           // process info parameters
      for (int i = optind; i < argc; ++i)
      {
        string param = lowercase(argv[i]);
        if (param == "all")
          for (int ip = MeInfo; ip <= SignalInfo; ++ip)
            printInfo((InfoParameter)ip);
        else if (param == "me")
          printInfo(MeInfo);
        else if (param == "fun")
          printInfo(FunctionalityInfo);
        else if (param == "op")
          printInfo(OperatorInfo);
        else if (param == "currop")
          printInfo(CurrentOperatorInfo);
        else if (param == "flstat")
          printInfo(FacilityLockStateInfo);
        else if (param == "flcap")
          printInfo(FacilityLockCapabilityInfo);
        else if (param == "pw")
          printInfo(PasswordInfo);
        else if (param == "pin")
          printInfo(PINInfo);
        else if (param == "clip")
          printInfo(CLIPInfo);
        else if (param == "forw")
          printInfo(CallForwardingInfo);
        else if (param == "batt")
          printInfo(BatteryInfo);
        else if (param == "biterr")
          printInfo(BitErrorInfo);
        else if (param == "sig")
          printInfo(SignalInfo);
        else if (param == "sca")
          printInfo(SCAInfo);
        else if (param == "cset")
          printInfo(CharSetInfo);
        else
          throw GsmException(
            stringPrintf(_("unknown information parameter '%s'"),
                         param.c_str()),
            ParameterError);
      }
    }
    else
    {                           // process operation
      operation = lowercase(operation);
      if (operation == "dial")
      {
        // dial: number
        checkParamCount(optind, argc, 1, 1);

        m->dial(argv[optind]);
        
        // wait for keypress from stdin
        char c;
        read(1, &c, 1);
      }
      else if (operation == "on")
      {
	  m->setFunctionalityLevel(1);
      }
      else if (operation == "off")
      {
	  m->setFunctionalityLevel(0);
      }
      else if (operation == "pin")
      {
        // pin: PIN
        checkParamCount(optind, argc, 1, 1);

        m->setPIN(argv[optind]);
      }
      else if (operation == "setop")
      {
        // setop: opmode numeric FIXME allow long and numeric too
        checkParamCount(optind, argc, 2, 2);
        string opmodeS = lowercase(argv[optind]);
        OPModes opmode;
        if (opmodeS == "automatic")
          opmode = AutomaticOPMode;
        else if (opmodeS == "manual")
          opmode = ManualOPMode;
        else if (opmodeS == "deregister")
          opmode = DeregisterOPMode;
        else if (opmodeS == "manualautomatic")
          opmode = ManualAutomaticOPMode;
        else
          throw GsmException(stringPrintf(_("unknown opmode parameter '%s'"),
                                          opmodeS.c_str()), ParameterError);

        m->setCurrentOPInfo(opmode, "" , "", checkNumber(argv[optind + 1]));
      }
      else if (operation == "lock")
      {
        // lock: facility [facilityclass] [passwd]
        checkParamCount(optind, argc, 1, 3);
        string passwd = (argc - optind == 3) ?
          (string)argv[optind + 2] : (string)"";
        
        m->lockFacility(argv[optind],
                        (argc - optind >= 2) ?
                        strToFacilityClass(argv[optind + 1]) :
                        (FacilityClass)ALL_FACILITIES,
                        passwd);
      }
      else if (operation == "unlock")
      {
        // unlock: facility [facilityclass] [passwd]
        checkParamCount(optind, argc, 1, 3);
        string passwd = argc - optind == 3 ? argv[optind + 2] : "";
        
        m->unlockFacility(argv[optind],
                          (argc - optind >= 2) ?
                          strToFacilityClass(argv[optind + 1]) :
                          (FacilityClass)ALL_FACILITIES,
                          passwd);
      }
      else if (operation == "setpw")
      {
        // set password: facility oldpasswd newpasswd
        checkParamCount(optind, argc, 1, 3);
        string oldPasswd = argc - optind >= 2 ? argv[optind + 1] : "";
        string newPasswd = argc - optind == 3 ? argv[optind + 2] : "";

        m->setPassword(argv[optind], oldPasswd, newPasswd);
      }
      else if (operation == "forw")
      {
        // call forwarding: mode reason number [facilityclass] [forwardtime]
        checkParamCount(optind, argc, 2, 5);

        // get optional parameters facility class and forwardtime
        int forwardTime = argc - optind == 5 ? checkNumber(argv[optind + 4]) :
          NOT_SET;
        FacilityClass facilityClass =
          argc - optind >= 4 ? strToFacilityClass(argv[optind + 3]) :
          (FacilityClass)ALL_FACILITIES;
        
        // get forward reason
        string reasonS = lowercase(argv[optind + 1]);
        ForwardReason reason;
        if (reasonS == "unconditional")
          reason = UnconditionalReason;
        else if (reasonS == "mobilebusy")
          reason = MobileBusyReason;
        else if (reasonS == "noreply")
          reason = NoReplyReason;
        else if (reasonS == "notreachable")
          reason = NotReachableReason;
        else if (reasonS == "all")
          reason = AllReasons;
        else if (reasonS == "allconditional")
          reason = AllConditionalReasons;
        else
          throw GsmException(
            stringPrintf(_("unknown forward reason parameter '%s'"),
                         reasonS.c_str()), ParameterError);
        
        // get mode
        string modeS = lowercase(argv[optind]);
        ForwardMode mode;
        if (modeS == "disable")
          mode = DisableMode;
        else if (modeS == "enable")
          mode = EnableMode;
        else if (modeS == "register")
          mode = RegistrationMode;
        else if (modeS == "erase")
          mode = ErasureMode;
        else
          throw GsmException(
            stringPrintf(_("unknown forward mode parameter '%s'"),
                         modeS.c_str()), ParameterError);

        m->setCallForwarding(reason, mode,
                             (argc - optind >= 3) ? argv[optind + 2] : "",
                             "", // subaddr
                             facilityClass, forwardTime);
      }
      else if (operation == "setsca")
      {
        // set sca: number
        checkParamCount(optind, argc, 1, 1);
        m->setServiceCentreAddress(argv[optind]);
      }
      else if (operation == "cset")
      {
        // set charset: string
        checkParamCount(optind, argc, 1, 1);
        m->setCharSet(argv[optind]);
      }
      else
         throw GsmException(stringPrintf(_("unknown operation '%s'"),
                                         operation.c_str()), ParameterError);
    }
  }
  catch (GsmException &ge)
  {
    cerr << argv[0] << _("[ERROR]: ") << ge.what() << endl;
    return 1;
  }
  return 0;
}
